{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# find cluster predictive regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from sklearn.svm import LinearSVC\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.datasets import load_iris\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "from collections import Counter, defaultdict\n",
    "import pickle\n",
    "\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load top 1k masked probes\n",
    "df_top_v1_prb_crc = pd.read_csv(\"./df_tcga_top1k_v1_probes_crc_pt2mask.csv\")\n",
    "df_top_v1_prb_crc.set_index(\"Unnamed: 0\", inplace=True)\n",
    "#print(df_top_v1_prb_crc.head())\n",
    "\n",
    "#load consensus cluster labels\n",
    "def create_clust_labels(file, data):\n",
    "    df_clusters = pd.read_csv(file, header=None)\n",
    "    cluster_dict = {}\n",
    "    for i, row in df_clusters.iterrows():\n",
    "        cluster_dict[row[0]] = row[1]\n",
    "    clust_labels = [cluster_dict[sam] for sam in data.index]\n",
    "    clust_labels = np.array(clust_labels).astype(str)\n",
    "    print(Counter(clust_labels))\n",
    "    return clust_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot feature importance for cluster generation\n",
    "def select_features(data, cluster_labels, outfile=None, prev_feat_to_keep = None, nseeds=10):\n",
    "    ''' generte an l1 regularized logreg that predicts cluster membership. \n",
    "    keep any features where the coeficient is not zero. \n",
    "    each coefficient can be interpreted as the expected change in log odds of cluster membership \n",
    "    for a one-unit increase in the percent methylation.\n",
    "    #https://rpubs.com/OmaymaS/182726\n",
    "    \n",
    "    output features and an annotation of which feature is hypermethylated has a positive non-zero coeff. \n",
    "    In the end there should be N sets of features, each \n",
    "    characterizing a co-methylation group for each methylation subtype\n",
    "    \n",
    "    Return:\n",
    "        signatures: a list of lists. each sublist a set of features whose hypermethylation defines\n",
    "        one methylation subtype\n",
    "    '''\n",
    "    classes = set(clust_labels)\n",
    "    print(f\"N clusters: {len(classes)}\")\n",
    "    \n",
    "    # MN LogReg regularized coefficients . . this works better than RF feature importances\n",
    "    ## NOTE: this method outputs 80-100 features depending on random seed so repeat many times\n",
    "    clf = LogisticRegression(penalty=\"l1\", solver='liblinear') \n",
    "    coefs=[]\n",
    "    for rep in range(nseeds):\n",
    "        clf.fit(data, clust_labels)\n",
    "        coefs.append(list(clf.coef_[0]))\n",
    "    df_coefs = pd.DataFrame.from_records(coefs)\n",
    "    #print(df_coefs.shape)\n",
    "    df_coefs.columns = data.columns\n",
    "    avg_coefs = df_coefs.apply(np.median, axis=0)\n",
    "    \n",
    "    signatures = []\n",
    "    if len(classes)==2:\n",
    "        plt.figure(figsize=(30,10))\n",
    "        plt.scatter(x=range(len(data.columns)), y=np.sort(avg_coefs))\n",
    "        cl1_feat_bool = [c > 0 for c in avg_coefs]\n",
    "        print(f\"cluster 1 features: {np.sum(np.array(cl1_feat_bool))}\")\n",
    "        cl2_feat_bool = [c < 0 for c in avg_coefs]\n",
    "        print(f\"cluster 2 features: {np.sum(np.array(cl2_feat_bool))}\")\n",
    "        signatures = [cl1_feat_bool, cl2_feat_bool]\n",
    "    else:\n",
    "        print(\"fix multinomial reporting\")\n",
    "        return\n",
    "        plt.figure()\n",
    "        for i, cluster in enumerate(classes):\n",
    "            plt.subplot(len(classes), 1, i+1)\n",
    "            plt.title(f\"Class {cluster} vs. all\")\n",
    "            plt.scatter(x=range(len(data.columns)), y=np.sort(clf.coef_[i,:]))\n",
    "            cl_feat_bool = [c > 0 for c in clf.coef_[i]]\n",
    "            print(f\"cluster {i} features: {np.sum(np.array(cl_feat_bool))}\")\n",
    "            signatures.append(cl_feat_bool)\n",
    "    \n",
    "    all_feat_to_keep = []\n",
    "    clust_annot = []\n",
    "    for i, sig in enumerate(signatures):\n",
    "        all_feat_to_keep.extend(np.array(data.columns)[sig])\n",
    "        clust_annot.extend(str(i)*len(np.array(data.columns)[sig])) \n",
    "\n",
    "    # from this we find that the majority of features are repeatedly selected with diff random seeds.\n",
    "    if prev_feat_to_keep:\n",
    "        print(f\"repeat features: {np.sum(np.array([x in prev_feat_to_keep for x in all_feat_to_keep]))}\")\n",
    "    \n",
    "    print(f\"features to keep: {len(all_feat_to_keep)}\")\n",
    "    df_out = pd.DataFrame(zip(all_feat_to_keep, clust_annot))\n",
    "    print(df_out.groupby(1).count())\n",
    "    df_out.columns = [\"marker\",\"cluster\"]\n",
    "    if outfile:\n",
    "        df_out.to_csv(f\"./{outfile}\", index=None)\n",
    "    return df_out, all_feat_to_keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'1': 213, '2': 148})\n",
      "N clusters: 2\n",
      "cluster 1 features: 24\n",
      "cluster 2 features: 62\n",
      "repeat features: 86\n",
      "features to keep: 86\n",
      "    0\n",
      "1    \n",
      "0  24\n",
      "1  62\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABrcAAAI/CAYAAADdp4JAAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzdf4is2X3f+c+Z6rK3JoHUbDQk7rJn5WVNZRlupI6HrMD/5JddTkByM3ZkL4J1ll30T0z+WFGgZoU1CoZRKER2lw0JwglJQGyiOE3tiDWUk9ghkDCLr6id9I7jIlpDZFV7k4mjMmxUSVqtZ/+4t2/6zu2+t/t2dT39VL1ecJmu89T0+QqGGdCbc55SVVUAAAAAAACgCV6oewAAAAAAAAC4KnELAAAAAACAxhC3AAAAAAAAaAxxCwAAAAAAgMYQtwAAAAAAAGgMcQsAAAAAAIDG2Kl7gMt84AMfqD74wQ/WPQYAAAAAAABr9tWvfvVfV1X18kXP7mzc+uAHP5j79+/XPQYAAAAAAABrVkr5F5c9cy0hAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI2xU/cAAAAAAAAAm2o8nWc0meV4scxut5PhoJ/9vV7dYzWauAUAAAAAAHALxtN5Dg6Psjw5TZLMF8scHB4licB1A64lBAAAAAAAuAWjyexR2DqzPDnNaDKraaLNIG4BAAAAAADcguPF8lrrXI24BQAAAAAAcAt2u51rrXM14hYAAAAAAMAtGA766bRbj6112q0MB/2aJtoMO3UPAAAAAAAAsIn293pJHrx763ixzG63k+Gg/2id5yNuAQAAAAAA3JL9vZ6YtWKuJQQAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaYyVxq5Ty10op/6qU8n9f8ryUUv6XUsrXSin/tJTyh1axLwAAAAAAANtlVSe3/nqSH33K8z+Z5Ace/vlkkr+8on0BAAAAAADYIiuJW1VV/aMk/+YpX/mxJH+zeuDtJN1SyvesYm8AAAAAAAC2x7reudVL8pvnPn/j4RoAAAAAAABc2bri1pWUUj5ZSrlfSrn/3nvv1T0OAAAAAAAAd8y64tY8yfed+/y9D9ceU1XVF6uqeq2qqtdefvnlNY0GAAAAAABAU6wrbr2V5L8pD3wkye9UVfVba9obAAAAAACADbGzil9SSvnfkvyRJB8opXwjyWeTtJOkqqq/kuQXk/ypJF9L8q0k/+0q9gUAAAAAAGC7rCRuVVX1Xz/jeZXkz65iLwAAAAAAALbXuq4lBAAAAAAAgBsTtwAAAAAAAGgMcQsAAAAAAIDGELcAAAAAAABoDHELAAAAAACAxhC3AAAAAAAAaAxxCwAAAAAAgMYQtwAAAAAAAGgMcQsAAAAAAIDGELcAAAAAAABoDHELAAAAAACAxhC3AAAAAAAAaAxxCwAAAAAAgMYQtwAAAAAAAGgMcQsAAAAAAIDGELcAAAAAAABoDHELAAAAAACAxhC3AAAAAAAAaAxxCwAAAAAAgMYQtwAAAAAAAGgMcQsAAAAAAIDGELcAAAAAAABoDHELAAAAAACAxhC3AAAAAAAAaAxxCwAAAAAAgMYQtwAAAAAAAGgMcQsAAAAAAIDG2Kl7AAAAAAAAgE00ns4zmsxyvFhmt9vJcNDP/l6v7rEaT9wCAAAAAABYsfF0noPDoyxPTpMk88UyB4dHSSJw3ZBrCQEAAAAAAFZsNJk9CltnlienGU1mNU20OcQtAAAAAACAFTteLK+1ztWJWwAAAAAAACu22+1ca52rE7cAAAAAAABWbDjop9NuPbbWabcyHPRrmmhz7NQ9AAAAAAAAwKbZ3+slefDurePFMrvdToaD/qN1np+4BQAAAAAAcAv293pi1i1wLSEAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGDt1DwAAAAAAALBJxtN5RpNZjhfL7HY7GQ762d/r1T3WxhC3AAAAAAAAVmQ8nefg8CjLk9MkyXyxzMHhUZIIXCviWkIAAAAAAIAVGU1mj8LWmeXJaUaTWU0TbR5xCwAAAAAAYEWOF8trrXN94hYAAAAAAMCK7HY711rn+sQtAAAAAACAFRkO+um0W4+tddqtDAf9mibaPDt1DwAAAAAAALAp9vd6SR68e+t4scxut5PhoP9onZsTtwAAAAAAAFZof68nZt0i1xICAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI2xU/cAAAAAAAAAm2I8nWc0meV4scxut5PhoJ/9vV7dY20UcQsAAAAAAGAFxtN5Dg6Psjw5TZLMF8scHB4licC1Qq4lBAAAAAAAWIHRZPYobJ1ZnpxmNJnVNNFmErcAAAAAAABW4HixvNY6z0fcAgAAAAAAWIHdbuda6zwfcQsAAAAAAGAFhoN+Ou3WY2uddivDQb+miTbTTt0DAAAAAAAAbIL9vV6SB+/eOl4ss9vtZDjoP1pnNcQtAAAAAACAFdnf64lZt0zcAgAAAAAAuKHxdO7E1pqIWwAAAAAAADcwns5zcHiU5clpkmS+WObg8ChJBK5b8ELdAwAAAAAAADTZaDJ7FLbOLE9OM5rMappos4lbAAAAAAAAN3C8WF5rnZsRtwAAAAAAAG5gt9u51jo3I24BAAAAAADcwHDQT6fdemyt025lOOjXNNFm26l7AAAAAAAAgCbb3+slefDurePFMrvdToaD/qN1VkvcAgAAAAAAuKH9vZ6YtSauJQQAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGmOn7gEAAAAAAACabDydZzSZ5XixzG63k+Ggn/29Xt1jbSxxCwAAAAAA4DmNp/McHB5leXKaJJkvljk4PEoSgeuWuJYQAAAAAADgOY0ms0dh68zy5DSjyaymiTafuAUAAAAAAPCcjhfLa61zc+IWAAAAAADAc9rtdq61zs2JWwAAAAAAAM9pOOin0249ttZptzIc9GuaaPPt1D0AAAAAAABAU+3v9ZI8ePfW8WKZ3W4nw0H/0Tqrt5K4VUr50ST/c5JWkp+vqurz73v+Z5KMkswfLv2vVVX9/Cr2BgAAAAAAqNP+Xk/MWqMbx61SSivJX0ryw0m+keRXSylvVVX1a+/76t+uqupnbrofAAAAAAAA22sV79z6w0m+VlXVb1RV9R+S/K0kP7aC3wsAAAAAAACPWUXc6iX5zXOfv/Fw7f1+vJTyT0spv1BK+b4V7AsAAAAAAMCWWUXcuoqvJPlgVVV/MMnfS/I3LvpSKeWTpZT7pZT777333ppGAwAAAAAAoClWEbfmSc6fxPreh2uPVFX121VV/fuHH38+yQ9e9IuqqvpiVVWvVVX12ssvv7yC0QAAAAAAANgkq4hbv5rkB0op319K+a4kP5XkrfNfKKV8z7mPH0vyz1awLwAAAAAAAFtm56a/oKqqb5dSfibJJEkryV+rqurdUsqfT3K/qqq3kvy5UsrHknw7yb9J8mduui8AAAAAAADbp1RVVfcMF3rttdeq+/fv1z0GAAAAAAAAa1ZK+WpVVa9d9GwV1xICAAAAAADAWohbAAAAAAAANIa4BQAAAAAAQGOIWwAAAAAAADSGuAUAAAAAAEBjiFsAAAAAAAA0hrgFAAAAAABAY4hbAAAAAAAANIa4BQAAAAAAQGOIWwAAAAAAADTGTt0DAAAAAAAANNF4Os9oMsvxYpndbifDQT/7e726x9p44hYAAAAAAMA1jafzHBweZXlymiSZL5Y5ODxKEoHrlrmWEAAAAAAA4JpGk9mjsHVmeXKa0WRW00TbQ9wCAAAAAAC4puPF8lrrrI64BQAAAAAAcE273c611lkdcQsAAAAAAOCahoN+Ou3WY2uddivDQb+mibbHTt0DAAAAAAAANM3+Xi/Jg3dvHS+W2e12Mhz0H61ze8QtAAAAAACA57C/1xOzauBaQgAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaQ9wCAAAAAACgMcQtAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABpD3AIAAAAAAKAxxC0AAAAAAAAaY6fuAQAAAAAAAJpmPJ1nNJnleLHMbreT4aCf/b1e3WNtBXELAAAAAADgGsbTeQ4Oj7I8OU2SzBfLHBweJYnAtQauJQQAAAAAALiG0WT2KGydWZ6cZjSZ1TTRdhG3AAAAAAAAruF4sbzWOqslbgEAAAAAAFzDbrdzrXVWS9wCAAAAAAC4ovF0nn/777/9xHqn3cpw0K9hou2zU/cAAAAAAAAATTCeznNwePTE+7ZeerGdz3701ezv9WqabLs4uQUAAAAAAHAFo8nsibCVJC9+146wtUbiFgAAAAAAwBUcL5bXWud2iFsAAAAAAABXsNvtXGud2yFuAQAAAAAAXMFw0E+n3XpsrdNuZTjo1zTRdhK3AAAAAAAArui7d/5jWnnpxXbefP2e922t2U7dAwAAAAAAANx14+k8B4dHWZ6cPlr7dyffqXGi7eXkFgAAAAAAwDOMJrPHwlaSLE9OM5rMappoe4lbAAAAAAAATzGezjNfLC98dnzJOrdH3AIAAAAAALjE2XWEl9ntdtY4DYm4BQAAAAAAcKmLriM802m3Mhz01zwR4hYAAAAAAMAlnnbt4Juv38v+Xm+N05CIWwAAAAAAAJe67NrBXrcjbNVE3AIAAAAAALjEcNBPp916bM11hPXaqXsAAAAAAACAu+rsdNZoMsvxYpndbifDQd+prRqJWwAAAAAAAE+xv9cTs+4QcQsAAAAAAOAS4+ncqa07RtwCAAAAAAC4wHg6z8HhUZYnp0mS+WKZg8OjJBG4avRC3QMAAAAAAADcRaPJ7FHYOrM8Oc1oMqtpIhJxCwAAAAAA4ELHi+W11lkPcQsAAAAAAOACu93OtdZZD3ELAAAAAADgAsNBP51267G1TruV4aBf00QkyU7dAwAAAAAAANxF+3u9JA/evXW8WGa328lw0H+0Tj3ELQAAAAAAgEvs7/XErDvGtYQAAAAAAAA0hpNbAAAAAAAA7zOezl1HeEeJWwAAAAAAAOeMp/McHB5leXKaJJkvljk4PEoSgesOcC0hAAAAAADAOaPJ7FHYOrM8Oc1oMqtpIs4TtwAAAAAAAM45Xiyvtc56iVsAAAAAAADn7HY711pnvcQtAAAAAACAc/7oH3g55X1rnXYrw0G/lnl4nLgFAAAAAADw0Hg6z9/96jzVubWS5Md/sJf9vV5dY3GOuAUAAAAAAPDQaDLL8uT0sbUqya/8+nv1DMQTxC0AAAAAAIA8OLU1XywvfHZ8yTrrt1P3AAAAAAAAAHX7zPgoX3r765c+3+121jgNT+PkFgAAAAAAsNXG03m+9PbXH3vP1nmddivDQX+tM3E5cQsAAAAAANhqo8ns0rCVJG++fi/7e721zcPTiVsAAAAAAMBWe9r7tHrdjrB1x4hbAAAAAADAVrvsfVolcR3hHSRuAQAAAAAAW2046KfTbj22VpJ84iOvOLV1B+3UPQAAAAAAAEBdxtN5RpNZlienaZWS06pKr9vJcNAXtu4ocQsAAAAAANhK4+k8B4dHWZ6cJklOqyqddkvYuuNcSwgAAAAAAGyd8XSeT335nUdh68zy5DSjyaymqbgKcQsAAAAAANgqZye2TqvqwufHi+WaJ+I6xC0AAAAAAGCrnL1j6zK73c4ap+G6xC0AAAAAAGCrPO1k1tk7t7i7xC0AAAAAAGCrdF9sX7jeKiVvvn4v+3u9NU/EdYhbAAAAAADAVhhP5/nw534p3/zWyRPP2q2SL3z8Q8JWA+zUPQAAAAAAAMBtG0/nOTg8uvRdW7/ru3aErYZwcgsAAAAAANh4o8ns0rCVJL+zfPI0F3eTuAUAAAAAAGy848Xyqc93u501TcJNiVsAAAAAAMDGe1q86rRbGQ76a5yGmxC3AAAAAACAjfdH/8DLKResv/RiO2++fs/7thpkp+4BAAAAAAAAbtN4Os/f/eo81bm1kuQTH3klP7d/r66xeE7iFgAAAAAAsLHG03k+9eV3clpVj61XSX7l19+rZyhuxLWEAAAAAADARhpP5zk4PHoibJ05XizXPBGrIG4BAAAAAAAb6XNfeTfLk9NLn+92O2uchlURtwAAAAAAgI0zns7zzW+dXPq8025lOOivcSJWRdwCAAAAAAA2zmgyu/RZq5S8+fq97O/11jgRqyJuAQAAAAAAG+dp79P6wsc/JGw1mLgFAAAAAABsnMvep9XttIWthhO3AAAAAACAjTMc9NNptx5b67RbeeNjr9Y0EauyU/cAAAAAAAAAq3Z2Oms0meV4scxut5PhoO/U1gYQtwAAAAAAgI20v9cTszaQawkBAAAAAABoDHELAAAAAACAxlhJ3Cql/GgpZVZK+Vop5dMXPP/uUsrffvj8/yylfHAV+wIAAAAAALBdbvzOrVJKK8lfSvLDSb6R5FdLKW9VVfVr57723yX5ZlVV/0Up5aeS/IUkP3nTvaEO4+k8o8ks88UyrVJyWlUpSaoaZnmhJN+pUtv+d2kW+2/3/ndpFvv7Z8H+9r9rs9h/u/e/S7PY3z8L9rf/XZvF/tu9/12axf7+WbD/7e//0ovtfPajr3r31gZZxcmtP5zka1VV/UZVVf8hyd9K8mPv+86PJfkbD3/+hSR/vJRSVrA3rNV4Os/B4VHmi2WS5LR68K/Zuv7D+52HG9f9H/6k/lnsv937n1f3LPavd//z6p7F/tu9/3l1z2L/7d7/vLpnsX+9+59X9yz23+79z6t7Fvtv9/7n1T2L/evd/7y6Z7H/7e3/zW+dZPgL72Q8nd/Cb6cONz65laSX5DfPff5Gkv/qsu9UVfXtUsrvJPm9Sf71CvaHW3f+tBYAAAAAAM1yclplNJk5vbUhVhG3VqaU8skkn0ySV155peZp2HaCFgAAAADA5jj2//VujFXErXmS7zv3+Xsfrl30nW+UUnaS/J4kv/3+X1RV1ReTfDFJXnvttbtwEpUt9ZnxUb709tfvxHFoAAAAAABubrfbqXsEVmQV79z61SQ/UEr5/lLKdyX5qSRvve87byX56Yc//0SSX66qSjfgThpP58IWAAAAAMAGabdKhoN+3WOwIjeOW1VVfTvJzySZJPlnSb5cVdW7pZQ/X0r52MOv/dUkv7eU8rUk/0OST990X7gto8ns2mGr3Mokz/ZCqXf/8+qexf7bvf95dc9i/3r3P6/uWey/3fufV/cs9t/u/c+rexb717v/eXXPYv/t3v+8umex/3bvf17ds9i/3v3Pq3sW+9/e/i+92M7oJz7kfVsbZCXv3Kqq6heT/OL71n723M//LsmfXsVecJvG0/m13rHVabfy5uv3/EsRAAAAAADWZCVxCzbBeDrPweHRlb/f63YyHPSFLQAAAAAAWCNxCx4aTWZZnpw+9TuCFgAAAAAA1EvcgoeOn3Id4f/0kx8WtAAAAAAA4A54oe4B4K7Y7XYuXO91O8IWAAAAAADcEU5usfXG03lGk1nmF5zc6rRbGQ76NUwFAAAAAABcRNxiq42n8xwcHl34rq2S5Md/sOfUFgAAAAAA3CGuJWSrjSazC8NWklRJfuXX31vvQAAAAAAAwFOJW2y14wuuIrzOcwAAAAAAYL3ELbbabrdzo+cAAAAAAMB6eecWW2c8neeNt97NYnny1O912q0MB/01TQUAAAAAAFyFuMVWuGrQOtPrdjIc9LO/17vlyQAAAAAAgOsQt9h44+k8w7/zTk6+U13p+71uJ//403/slqcCAAAAAACeh7jFRhtP5/nUl9/JaXW1sJUkx4vlLU4EAAAAAADchLjFxhlP5xlNZpk/Z6Ta7XZWPBEAAAAAALAq4hYbZTyd5+DwKMuT0+f6+zvtVoaD/oqnAgAAAAAAVkXcYqOMJrNrh60XSvKd6sG7toaDfvb3erc0HQAAAAAAcFPiFhvlqlcRvvRiO5/96KtCFgAAAAAANIy4xcYYT+cpSaqnfKck+Ys/+WFRCwAAAAAAGuqFugeAVRlNZs8MW5/4yCvCFgAAAAAANJiTW2yE8XT+1CsJvU8LAAAAAAA2g7hFY42n84wms2e+Z6vX7eQff/qPrWkqAAAAAADgNolbNNJ4Os/B4VGWJ6dP/V6n3cpw0F/TVAAAAAAAwG0Tt2ic8XSeT335nZxWT3vD1gNvvn7PVYQAAAAAALBBxC0a5TPjo3zp7a/n2VnrwXWEwhYAAAAAAGwWcYtGGE/neeOtd7NYnlzp+yVxHSEAAAAAAGwgcYs776rv1zpTknziI684tQUAAAAAABtI3OLOG01mVw5bvW4nw0Ff2AIAAAAAgA0lbnHnHS+WT31ekvzFn/ywoAUAAAAAAFvghboHgGfpvti+9JkrCAEAAAAAYLs4ucWdNZ7O88Zb72axPLnw+UsvtvPZj74qbAEAAAAAwBYRt7iTxtN5Dg6PLn3XVrfTzvRnf2TNUwEAAAAAAHVzLSF30ue+8u6lYStJfueS01wAAAAAAMBmE7e4U8bTeT78uV/KN7/19Hi12+2saSIAAAAAAOAucS0hd8Znxkf50ttfT/WM73XarQwH/bXMBAAAAAAA3C3iFnfCeDq/Uth66cV2PvvRV7O/11vLXAAAAAAAwN0ibnEnjCazZ4atbqed6c/+yFrmAQAAAAAA7ibv3OJOOF4sn/q8027ljY+9uqZpAAAAAACAu0rconbj6TwvlHLp85debOfN1++5ihAAAAAAAHAtIfUaT+c5ODzKafXkpYQlySc+8kp+bv/e+gcDAAAAAADuJHGLWn3uK+9meXL6xHqrlHzh4x9yWgsAAAAAAHiMawmpxXg6z4c/90v55rdOLnz+naoStgAAAAAAgCc4ucXafWZ8lC+9/fU8eRHhf7Tb7axtHgAAAAAAoDmc3GKtxtP5M8NWkgwH/bXMAwAAAAAANIu4xdqMp/N86svvPDNsdTttVxICAAAAAAAXci0ha3GVqwiTpNNu5Y2PvbqWmQAAAAAAgOYRt7h1V72K8KUX2/nsR191agsAAAAAALiUuMWtG01mTw1bJcknPvJKfm7/3rpGAgAAAAAAGkrc4tYdL5aXPmuVki98/ENOawEAAAAAAFfyQt0DsPl2u50L10sibAEAAAAAANcibnHrhoN+Ou3WY2tnVxEKWwAAAAAAwHW4lpBbM57O88Zb72axPEmSvFCS71RJr9vJcNAXtgAAAAAAgGsTt1i590etM9+pknarCFsAAAAAAMBzcy0hKzWeznNwePRE2DpzclplNJmteSoAAAAAAGBTiFus1Ggyy/Lk9KnfOV4s1zQNAAAAAACwacQtVuoq4Wq321nDJAAAAAAAwCYSt1ipZ4Wrs3duAQAAAAAAPA9xi5UaDvrptFsXPnvpxXZGP/Gh7O/11jwVAAAAAACwKXbqHoDNchauRpNZjhfL7HY7GQ76ghYAAAAAALAS4hYrt7/XE7MAAAAAAIBb4VpCAAAAAAAAGkPcAgAAAAAAoDHELQAAAAAAABrDO7dYqfF0ntFkluPFMrvdToaDvvdvAQAAAAAAKyNusTLj6TwHh0dZnpwmSeaLZQ4Oj5JE4AIAAAAAAFbCtYSsxHg6z6e+/M6jsHVmeXKa0WRW01QAAAAAAMCmEbe4sbMTW6dVdeHz48VyzRMBAAAAAACbStzixkaT2RMnts7b7XbWOA0AAAAAALDJxC1ubP6Uk1mddivDQX+N0wAAAAAAAJtM3OJGxtN5yiXPWqXkzdfvZX+vt9aZAAAAAACAzSVucSOjySwXvWmrJPnCxz8kbAEAAAAAACslbnEjx5dcSVglwhYAAAAAALBy4hY3stvtXLjeu2QdAAAAAADgJsQtbmQ46KfTbj221mm3Mhz0a5oIAAAAAADYZDt1D0BzjafzjCazLE9O0yolp1WVXreT4aDvSkIAAAAAAOBWiFtc23g6zxtvvZvF8uTR2mlVPTqxJWwBAAAAAAC3xbWEXMt4Os/B4dFjYevM8uQ0o8mshqkAAAAAAIBtIW5xLWfXEF7meLFc4zQAAAAAAMC2cS0hV3L2fq35M+LVbrezpokAAAAAAIBtJG7xTGdXET7txGeqvLMAABPwSURBVFaSR+/cAgAAAAAAuC3iFs/0ua+8+8yw9dKL7Xz2o69mf6+3pqkAAAAAAIBtJG5xqfF0njfeejeL5ckzvzv92R9Zw0QAAAAAAMC2E7e40FWvIkySnvdsAQAAAAAAa/JC3QNwN40msyuFLe/ZAgAAAAAA1snJLS50vFg+8zu9bifDQd97tgAAAAAAgLURt7jQbreT+SWBq9Nu5c3X74laAAAAAADA2rmWkAsNB/102q0n1l96sS1sAQAAAAAAtXFyiwudxavRZJbjxTK7riAEAAAAAADuAHGLS+3v9cQsAAAAAADgTnEtIQAAAAAAAI0hbgEAAAAAANAY4hYAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjbFT9wDcLePpPKPJLPPFMq1SclpV6XU7GQ762d/r1T0eAAAAAACw5cQtHvnM+ChfevvrqR5+Pq0e/DRfLHNweJQkAhcAAAAAAFAr1xKS5MGJrfNh6/2WJ6cZTWZrnQkAAAAAAOD9xC2SJKPJ7NKwdeZ4sVzLLAAAAAAAAJcRt0hytXC12+2sYRIAAAAAAIDLiVskeXa46rRbGQ76a5oGAAAAAADgYuIWSZLhoJ9Ou3Xhs163kzdfv5f9vd6apwIAAAAAAHjcTt0DcDechavRZJbjxTK73U6Gg76gBQAAAAAA3CniFo/s7/XELAAAAAAA4E5zLSEAAAAAAACNIW4BAAAAAADQGOIWAAAAAAAAjSFuAQAAAAAA0BjiFgAAAAAAAI2xU/cA1G88nWc0meV4scxut5PhoJ/9vV7dYwEAAAAAADxB3Npy4+k8B4dHWZ6cJknmi2UODo+SROACAAAAAADuHNcSbrnRZPYobJ1ZnpxmNJnVNBEAAAAAAMDlxK0td7xYXmsdAAAAAACgTuLWltvtdq61DgAAAAAAUKcbxa1Syn9aSvl7pZR//vCvL13yvdNSyv/18M9bN9mT1RoO+um0W4+tddqtDAf9miYCAAAAAAC43E1Pbn06yT+oquoHkvyDh58vsqyq6sMP/3zshnuyQvt7vbz5+r30up2UJL1uJ2++fi/7e726RwMAAAAAAHhCqarq+f/mUmZJ/khVVb9VSvmeJP+wqqonjvyUUv6/qqp+93V+92uvvVbdv3//uWcDAAAAAACgmUopX62q6rWLnt305Nbvq6rqtx7+/P8m+X2XfO8/KaXcL6W8XUrZv+GeAAAAAAAAbKmdZ32hlPL3k/z+Cx79j+c/VFVVlVIuOwb2n1VVNS+l/OdJfrmUclRV1f9zwV6fTPLJJHnllVeeOTwAAAAAAADb5Zlxq6qqP3HZs1LKvyylfM+5awn/1SW/Y/7wr79RSvmHSfaSPBG3qqr6YpIvJg+uJbzS/wIAAAAAAAC2xk2vJXwryU8//Pmnk/zv7/9CKeWlUsp3P/z5A0l+KMmv3XBfVmQ8neeHPv/L+f5P/x/5oc//csbTed0jAQAAAAAAXOqmcevzSX64lPLPk/yJh59TSnmtlPLzD7/zXya5X0p5J8mvJPl8VVXi1h0wns5zcHiU+WKZKsl8sczB4ZHABQAAAAAA3FnPvJbwaaqq+u0kf/yC9ftJ/vuHP/+TJPdusg+3YzSZZXly+tja8uQ0o8ks+3u9mqYCAAAAAAC43E1PbtFQ4+k888XywmfHl6wDAAAAAADUTdzaQmfXEV5mt9tZ4zQAAAAAAABXJ25toYuuIzzTabcyHPTXPBEAAAAAAMDViFtb6GnXDr75+j3v2wIAAAAAAO4scWsLXXbtYK/bEbYAAAAAAIA7TdzaQsNBP51267E11xECAAAAAABNsFP3AKzf2ems0WSW48Uyu91OhoO+U1sAAAAAAMCdJ25tqf29npgFAAAAAAA0jmsJAQAAAPj/27vb2ErTsg7g/yudrhwwoSAGbZeVNW5qNhos2SBEYwgQCkjcQhAhGjeI4YtGNFjD6AfiB2JMVdRISAgvojEgWUvZKLFBJMEvEsBGC6wNm1VgW9506fp2lG65/XDODJ3dYZduZ3t6n/5+yWT6XM8zc65+uXJ3/vPcNwBAN4RbAAAAAAAAdEO4BQAAAAAAQDeEWwAAAAAAAHRDuAUAAAAAAEA3hFsAAAAAAAB0Q7gFAAAAAABAN4RbAAAAAAAAdEO4BQAAAAAAQDeEWwAAAAAAAHRDuAUAAAAAAEA3hFsAAAAAAAB0Q7gFAAAAAABAN4RbAAAAAAAAdEO4BQAAAAAAQDeEWwAAAAAAAHRDuAUAAAAAAEA3hFsAAAAAAAB0Q7gFAAAAAABAN4RbAAAAAAAAdEO4BQAAAAAAQDeEWwAAAAAAAHRDuAUAAAAAAEA3Lky6AU7fxtZu1jZ3src/zPzcIKvLi1lZWph0WwAAAAAAAA9LuHXObGzt5uL6doYHh0mS3f1hLq5vJ4mACwAAAAAAOPNsS3iObGzt5nXv/cfLwdYlw4PDrG3uTKgrAAAAAACAb51w65y49MbWYWtXvb+3PzzljgAAAAAAAI5PuHVOrG3uPOiNraPm5wan2A0AAAAAAMAjI9w6Jx7qzazB7ExWlxdPsRsAAAAAAIBHRrh1Tsw9dvaq9Zmq/NZLfzArSwun3BEAAAAAAMDxCbfOgY2t3fzX/97/oPrsTOV3X/40wRYAAAAAANAN4dY5sLa5k4OvtwfVH3fdBcEWAAAAAADQFeHWOfDNztu6b3hwyp0AAAAAAACcjHDrHJifGxyrDgAAAAAAcFYJt86B1eXFDGZnrqgNZmeyurw4oY4AAAAAAAAemQuTboBH36VztdY2d7K3P8z83CCry4vO2wIAAAAAALoj3DonVpYWhFkAAAAAAED3bEsIAAAAAABAN4RbAAAAAAAAdEO4BQAAAAAAQDecuTXlNrZ2s7a5k739YebnBlldXnT2FgAAAAAA0C3h1hTb2NrNxfXtDA8OkyS7+8NcXN9OEgEXAAAAAADQJdsSTrG1zZ3LwdYlw4PDrG3uTKgjAAAAAACAkxFuTbG9/eGx6gAAAAAAAGedcGuKzc8NjlUHAAAAAAA464RbU2x1eTGD2ZkraoPZmawuL06oIwAAAAAAgJO5MOkGePSsLC0kGZ29tbc/zPzcIKvLi5frAAAAAAAAvRFuTbmVpQVhFgAAAAAAMDVsSwgAAAAAAEA3hFsAAAAAAAB0Q7gFAAAAAABAN5y5NYU2tnaztrmT3f1hZqpy2FoW5gZZXV50/hYAAAAAANA14daU2djazcX17QwPDpMkh60lSXb3h7m4vp0kAi4AAAAAAKBbtiWcMmubO5eDrQcaHhxmbXPnlDsCAAAAAAC4doRbU2Zvf3ii+wAAAAAAAGeZcGvKzM8NTnQfAAAAAADgLBNuTZnV5cUMZmeuem8wO5PV5cVT7ggAAAAAAODauTDpBri2VpYWkozO3trdH2amKoetZWFukNXlxcv3AQAAAAAAeiTcmkIrSwtCLAAAAAAAYCrZlhAAAAAAAIBuCLcAAAAAAADohm0Jp8zG1m7WNneytz/MvHO2AAAAAACAKSPcmiIbW7u5uL6d4cFhkmR3f5iL69tJIuACAAAAAACmgm0Jp8ja5s7lYOuS4cFh1jZ3JtQRAAAAAADAtSXcmiJ7+8Nj1QEAAAAAAHoj3Joi83ODY9UBAAAAAAB6I9yaIqvLixnMzlxRG8zOZHV5cUIdAQAAAAAAXFsXJt0A187K0kKS0dlbe/vDzM8Nsrq8eLkOAAAAAADQO+HWlFlZWhBmAQAAAAAAU8u2hAAAAAAAAHTDm1tTYmNr13aEAAAAAADA1BNuTYGNrd1cXN/O8OAwSbK7P8zF9e0kEXABAAAAAABTxbaEU2Btc+dysHXJ8OAwa5s7E+oIAAAAAADg0SHcmgJ7+8Nj1QEAAAAAAHol3JoC83ODY9UBAAAAAAB6JdyaAqvLixnMzlxRG8zOZHV5cUIdAQAAAAAAPDouTLoBTm5laSHJ6Oytvf1h5ucGWV1evFwHAAAAAACYFsKtKbGytCDMAgAAAAAApp5tCQEAAAAAAOiGcAsAAAAAAIBuCLcAAAAAAADohnALAAAAAACAbgi3AAAAAAAA6IZwCwAAAAAAgG4ItwAAAAAAAOjGhUk3wMltbO1mbXMne/vDzM8Nsrq8mJWlhUm3BQAAAAAAcM0Jtzq3sbWbi+vbGR4cJkl294e5uL6dJAIuAAAAAABg6tiWsHNrmzuXg61LhgeHWdvcmVBHAAAAAAAAjx7hVuf29ofHqgMAAAAAAPRMuNW5+bnBseoAAAAAAAA9E251bnV5MYPZmStqg9mZrC4vTqgjAAAAAACAR8+FSTfAyawsLSQZnb21tz/M/Nwgq8uLl+sAAAAAAADTRLg1BVaWFoRZAAAAAADAuWBbQgAAAAAAALoh3AIAAAAAAKAbwi0AAAAAAAC6IdwCAAAAAACgG8ItAAAAAAAAuiHcAgAAAAAAoBvCLQAAAAAAALoh3AIAAAAAAKAbwi0AAAAAAAC6IdwCAAAAAACgG8ItAAAAAAAAunGicKuqfrKqPlVVX6+qWx7iuRdU1U5V3VVVrz/JZwIAAAAAAHB+nfTNrU8meWmSj3yzB6pqJsmbk7wwyc1JXllVN5/wcwEAAAAAADiHLpzkD7fW7kySqnqox56R5K7W2t3jZ9+T5NYknz7JZwMAAAAAAHD+nMaZWwtJPn/k+p5xDQAAAAAAAI7lYd/cqqq/SfJdV7n1G62191/LZqrqNUlekyQ33HDDtfyrAQAAAAAAmAIPG2611p53ws/YTfKUI9fXj2tX+6y3Jnlrktxyyy3thJ8LAAAAAADAlDmNbQk/luSmqrqxqq5L8ookd5zC5wIAAAAAADBlThRuVdVLquqeJM9K8ldVtTmuz1fVB5KktXZ/kl9MspnkziTvba196mRtAwAAAAAAcB497LaED6W19r4k77tKfS/Ji45cfyDJB07yWQAAAAAAAHAa2xICAAAAAADANVGttUn3cFVV9ZUkn510H515UpJ/m3QTAI+A+QX0zAwDemaGAb0yv4CemWHfmu9prX3n1W6c2XCL46uqj7fWbpl0HwDHZX4BPTPDgJ6ZYUCvzC+gZ2bYydmWEAAAAAAAgG4ItwAAAAAAAOiGcGu6vHXSDQA8QuYX0DMzDOiZGQb0yvwCemaGnZAztwAAAAAAAOiGN7cAAAAAAADohnBrClTVC6pqp6ruqqrXT7ofgAeqqqdU1Yer6tNV9amqeu24/sSq+mBVfWb8+xPG9aqqPxzPtX+qqqdP9jsAzruqmqmqrar6y/H1jVX10fGc+vOqum5c/7bx9V3j+0+dZN8AVTVXVbdX1T9X1Z1V9SxrMKAHVfUr458fP1lV766qx1iDAWdVVb2jqr5cVZ88Ujv2mquqbhs//5mqum0S30svhFudq6qZJG9O8sIkNyd5ZVXdPNmuAB7k/iSva63dnOSZSX5hPKten+RDrbWbknxofJ2MZtpN41+vSfKW028Z4AqvTXLnkevfTvKm1tr3JflqkleP669O8tVx/U3j5wAm6Q+S/HVr7fuTPC2jWWYNBpxpVbWQ5JeS3NJa+4EkM0leEWsw4Oz64yQveEDtWGuuqnpikjck+eEkz0jyhkuBGA8m3OrfM5Lc1Vq7u7X2tSTvSXLrhHsCuEJr7QuttX8Yf/2fGf2jykJG8+pd48felWRl/PWtSf6kjfx9krmq+u5TbhsgSVJV1yf58SRvG19XkuckuX38yAPn16W5dnuS546fBzh1VfX4JD+W5O1J0lr7WmttP9ZgQB8uJBlU1YUkj03yhViDAWdUa+0jSe59QPm4a67lJB9srd3bWvtqkg/mwYEZY8Kt/i0k+fyR63vGNYAzabw9xFKSjyZ5cmvtC+NbX0zy5PHXZhtwlvx+kl9L8vXx9Xck2W+t3T++PjqjLs+v8f37xs8DTMKNSb6S5J3jrVXfVlWPizUYcMa11naT/E6Sz2UUat2X5BOxBgP6ctw1l7XYMQi3ADg1VfXtSf4iyS+31v7j6L3WWkvSJtIYwDdRVS9O8uXW2icm3QvAI3AhydOTvKW1tpTkv/ON7XCSWIMBZ9N4G65bMwrp55M8Lt5eADpmzXXtCbf6t5vkKUeurx/XAM6UqprNKNj6s9ba+rj8pUtb3Yx///K4brYBZ8WPJPmJqvrXjLZ/fk5G59fMjbfISa6cUZfn1/j+45P8+2k2DHDEPUnuaa19dHx9e0ZhlzUYcNY9L8m/tNa+0lo7SLKe0brMGgzoyXHXXNZixyDc6t/HktxUVTdW1XUZHa55x4R7ArjCeK/ztye5s7X2e0du3ZHktvHXtyV5/5H6z9bIM5Pcd+Q1boBT01q72Fq7vrX21IzWWX/bWvvpJB9O8rLxYw+cX5fm2svGz/vfecBEtNa+mOTzVbU4Lj03yadjDQacfZ9L8syqeuz458lL88saDOjJcddcm0meX1VPGL/B+vxxjasoc75/VfWijM6CmEnyjtbaGyfcEsAVqupHk/xdku1848yaX8/o3K33JrkhyWeTvLy1du/4h5c/ymjbif9J8qrW2sdPvXGAI6rq2Ul+tbX24qr63oze5Hpikq0kP9Na+7+qekySP83obMF7k7yitXb3pHoGqKofSvK2JNcluTvJqzL6j67WYMCZVlW/meSnktyf0Xrr5zM6e8YaDDhzqurdSZ6d5ElJvpTkDUk2csw1V1X9XEb/ZpYkb2ytvfM0v4+eCLcAAAAAAADohm0JAQAAAAAA6IZwCwAAAAAAgG4ItwAAAAAAAOiGcAsAAAAAAIBuCLcAAAAAAADohnALAAAAAACAbgi3AAAAAAAA6IZwCwAAAAAAgG78P4s8QefqPIlNAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 2160x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#df_top_v1_prb_crc.head()\n",
    "clust_labels = create_clust_labels(\"./ConClust_kmeans_mask/ConClust_kmeans_mask.k=2.consensusClass.csv\", df_top_v1_prb_crc)\n",
    "df_feat_kmean_mask_k2, feat_to_keep = select_features(df_top_v1_prb_crc, clust_labels, \n",
    "                                                      outfile=\"./kmeans_mask_k2_probes.txt\", \n",
    "                                                      prev_feat_to_keep=feat_to_keep,\n",
    "                                                      nseeds=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter({'4': 107, '3': 94, '1': 88, '2': 72})\n",
      "N clusters: 4\n",
      "fix multinomial reporting\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/billiegould/Desktop/research/env/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.\n",
      "  FutureWarning)\n",
      "/Users/billiegould/Desktop/research/env/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:469: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.\n",
      "  \"this warning.\", FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "clust_labels = create_clust_labels(\"./ConClust_kmeans_mask/ConClust_kmeans_mask.k=4.consensusClass.csv\", df_top_v1_prb_crc)\n",
    "df_feat_kmean_mask_k4 = select_features(df_top_v1_prb_crc, clust_labels, outfile=\"./kmeans_mask_k4_probes.txt\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# identify v1 panel regions containing selected tcga markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(632108, 632228, 'chr1_632108_632228'), (935143, 935263, 'chr1_935143_935263'), (943939, 944095, 'chr1_943939_944095'), (963412, 963532, 'chr1_963412_963532'), (970916, 971158, 'chr1_970916_971158')]\n"
     ]
    }
   ],
   "source": [
    "# load v1 panel\n",
    "v1_16K_regions = defaultdict(list)\n",
    "with open(\"../methyl_panel_v1_target_twist.bed\", \"r\") as fin:\n",
    "    bed = pd.read_csv(fin, sep=\"\\t\", header=None)\n",
    "    #print(bed.head())\n",
    "    bed = bed.iloc[:,0:3]\n",
    "    bed.columns = ['chrom', 'start', 'stop']\n",
    "    for _, row in bed.iterrows():\n",
    "        reg_name = str(row.chrom) + \"_\"+ str(row.start) + \"_\"+ str(row.stop)\n",
    "        v1_16K_regions[row['chrom']].append((row['start'],row['stop'], reg_name))\n",
    "\n",
    "print(v1_16K_regions[\"chr1\"][0:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(9652346, 9652712, 'chr1_9652346_9652712'), (12063181, 12063620, 'chr1_12063181_12063620'), (34163174, 34163579, 'chr1_34163174_34163579'), (34930048, 34930228, 'chr1_34930048_34930228')]\n",
      "['chr1_9652346_9652712', 'chr1_12063181_12063620', 'chr1_34163174_34163579', 'chr1_34930048_34930228']\n"
     ]
    }
   ],
   "source": [
    "# load 434 regions bed\n",
    "# 434 Regions selected by early modeling and by lit search\n",
    "from core.gcloud.file import opener\n",
    "modeling_regions = defaultdict(list)\n",
    "modeling_regions_names = []\n",
    "with opener('gs://reference-datafiles/beds/freenome_v1_panel_manual_and_automated_per_region_hmf_definition_20190823.bed', 'r') as handle:\n",
    "    expected_bed = pd.read_csv(handle, sep=\"\\t\", header=None)\n",
    "    expected_bed.columns = ['chrom', 'start', 'stop']\n",
    "    for _, row in expected_bed.iterrows():\n",
    "        reg_name = str(row.chrom) + \"_\"+ str(row.start) + \"_\"+ str(row.stop)\n",
    "        modeling_regions[row['chrom']].append((row['start'],row['stop'], reg_name))\n",
    "        modeling_regions_names.append(reg_name)\n",
    "\n",
    "print(modeling_regions[\"chr1\"][0:4])\n",
    "print(modeling_regions_names[0:4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Gene_Symbol</th>\n",
       "      <th>Gene_Type</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Position_to_TSS</th>\n",
       "      <th>CGI_Coordinate</th>\n",
       "      <th>Feature_Type</th>\n",
       "      <th>in_v1_panel</th>\n",
       "      <th>in_v3_panel</th>\n",
       "      <th>v1_panel_region</th>\n",
       "      <th>v1_region</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Composite Element REF</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg00000029</th>\n",
       "      <td>chr16</td>\n",
       "      <td>53434200</td>\n",
       "      <td>53434201</td>\n",
       "      <td>RBL2;RBL2;RBL2</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding</td>\n",
       "      <td>ENST00000262133.9;ENST00000544405.5;ENST000005...</td>\n",
       "      <td>-221;-1420;222</td>\n",
       "      <td>CGI:chr16:53434489-53435297</td>\n",
       "      <td>N_Shore</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000108</th>\n",
       "      <td>chr3</td>\n",
       "      <td>37417715</td>\n",
       "      <td>37417716</td>\n",
       "      <td>C3orf35;C3orf35;C3orf35;C3orf35;C3orf35;C3orf3...</td>\n",
       "      <td>lincRNA;lincRNA;lincRNA;lincRNA;lincRNA;lincRN...</td>\n",
       "      <td>ENST00000328376.8;ENST00000332506.6;ENST000004...</td>\n",
       "      <td>18552;18552;6505;31445;18143;447;18552;18552</td>\n",
       "      <td>CGI:chr3:37451927-37453047</td>\n",
       "      <td>.</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000109</th>\n",
       "      <td>chr3</td>\n",
       "      <td>172198247</td>\n",
       "      <td>172198248</td>\n",
       "      <td>FNDC3B;FNDC3B;FNDC3B;FNDC3B;FNDC3B;FNDC3B</td>\n",
       "      <td>protein_coding;protein_coding;protein_coding;p...</td>\n",
       "      <td>ENST00000336824.7;ENST00000415807.5;ENST000004...</td>\n",
       "      <td>157692;158618;151333;71272;158587;71273</td>\n",
       "      <td>CGI:chr3:172039703-172040934</td>\n",
       "      <td>.</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000165</th>\n",
       "      <td>chr1</td>\n",
       "      <td>90729117</td>\n",
       "      <td>90729118</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>CGI:chr1:90724932-90727247</td>\n",
       "      <td>S_Shore</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg00000236</th>\n",
       "      <td>chr8</td>\n",
       "      <td>42405776</td>\n",
       "      <td>42405777</td>\n",
       "      <td>VDAC3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENST00000022615.7</td>\n",
       "      <td>13872</td>\n",
       "      <td>CGI:chr8:42410918-42411241</td>\n",
       "      <td>.</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Chromosome      Start        End  \\\n",
       "Composite Element REF                                    \n",
       "cg00000029                 chr16   53434200   53434201   \n",
       "cg00000108                  chr3   37417715   37417716   \n",
       "cg00000109                  chr3  172198247  172198248   \n",
       "cg00000165                  chr1   90729117   90729118   \n",
       "cg00000236                  chr8   42405776   42405777   \n",
       "\n",
       "                                                             Gene_Symbol  \\\n",
       "Composite Element REF                                                      \n",
       "cg00000029                                                RBL2;RBL2;RBL2   \n",
       "cg00000108             C3orf35;C3orf35;C3orf35;C3orf35;C3orf35;C3orf3...   \n",
       "cg00000109                     FNDC3B;FNDC3B;FNDC3B;FNDC3B;FNDC3B;FNDC3B   \n",
       "cg00000165                                                             .   \n",
       "cg00000236                                                         VDAC3   \n",
       "\n",
       "                                                               Gene_Type  \\\n",
       "Composite Element REF                                                      \n",
       "cg00000029                  protein_coding;protein_coding;protein_coding   \n",
       "cg00000108             lincRNA;lincRNA;lincRNA;lincRNA;lincRNA;lincRN...   \n",
       "cg00000109             protein_coding;protein_coding;protein_coding;p...   \n",
       "cg00000165                                                             .   \n",
       "cg00000236                                                protein_coding   \n",
       "\n",
       "                                                           Transcript_ID  \\\n",
       "Composite Element REF                                                      \n",
       "cg00000029             ENST00000262133.9;ENST00000544405.5;ENST000005...   \n",
       "cg00000108             ENST00000328376.8;ENST00000332506.6;ENST000004...   \n",
       "cg00000109             ENST00000336824.7;ENST00000415807.5;ENST000004...   \n",
       "cg00000165                                                             .   \n",
       "cg00000236                                             ENST00000022615.7   \n",
       "\n",
       "                                                    Position_to_TSS  \\\n",
       "Composite Element REF                                                 \n",
       "cg00000029                                           -221;-1420;222   \n",
       "cg00000108             18552;18552;6505;31445;18143;447;18552;18552   \n",
       "cg00000109                  157692;158618;151333;71272;158587;71273   \n",
       "cg00000165                                                        .   \n",
       "cg00000236                                                    13872   \n",
       "\n",
       "                                     CGI_Coordinate Feature_Type  in_v1_panel  \\\n",
       "Composite Element REF                                                           \n",
       "cg00000029              CGI:chr16:53434489-53435297      N_Shore        False   \n",
       "cg00000108               CGI:chr3:37451927-37453047            .        False   \n",
       "cg00000109             CGI:chr3:172039703-172040934            .        False   \n",
       "cg00000165               CGI:chr1:90724932-90727247      S_Shore        False   \n",
       "cg00000236               CGI:chr8:42410918-42411241            .        False   \n",
       "\n",
       "                       in_v3_panel v1_panel_region v1_region  \n",
       "Composite Element REF                                         \n",
       "cg00000029                   False              NA        NA  \n",
       "cg00000108                   False              NA        NA  \n",
       "cg00000109                   False              NA        NA  \n",
       "cg00000165                   False              NA        NA  \n",
       "cg00000236                   False              NA        NA  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load tcga probe metadata\n",
    "df_probe_metadata = pickle.load(open(\"./df_keep_probes_metadata.pkl\", \"rb\"))\n",
    "df_probe_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       marker  cluster\n",
      "0  cg03616722        0\n",
      "1  cg06509223        0\n",
      "2  cg24506221        0\n",
      "3  cg20553577        0\n",
      "4  cg16814680        0\n",
      "['cg03616722', 'cg06509223', 'cg24506221', 'cg20553577', 'cg16814680', 'cg02293936', 'cg06266993', 'cg23564700', 'cg01873886', 'cg02246055', 'cg16570157', 'cg17736443', 'cg02159489', 'cg22681934', 'cg08154437', 'cg08248516', 'cg24008544', 'cg23332582', 'cg14159026', 'cg03483654', 'cg14255824', 'cg01939274', 'cg26522157', 'cg17603288', 'cg24371075', 'cg25830696', 'cg27438798', 'cg15825786', 'cg13021333', 'cg18745566', 'cg02907374', 'cg06368118', 'cg14081924', 'cg18136062', 'cg18800085', 'cg17038667', 'cg20893717', 'cg15454726', 'cg04100696', 'cg00366818', 'cg10748160', 'cg04689080', 'cg16485558', 'cg25373100', 'cg16882226', 'cg12376068', 'cg11514698', 'cg06466031', 'cg22761670', 'cg11964564', 'cg08042975', 'cg20115218', 'cg05817517', 'cg21309100', 'cg24933645', 'cg05099909', 'cg11258164', 'cg05293775', 'cg17512364', 'cg13643796', 'cg10957242', 'cg25324047', 'cg20895877', 'cg18875716', 'cg23448584', 'cg04970352', 'cg22149516', 'cg26525127', 'cg07134065', 'cg21952901', 'cg09350341', 'cg08661899', 'cg13640626', 'cg25457956', 'cg25925764', 'cg20478129', 'cg07551545', 'cg15355118', 'cg12307787', 'cg21662160', 'cg25428398', 'cg19540689', 'cg25561581', 'cg26394244', 'cg09508531', 'cg09279240']\n"
     ]
    }
   ],
   "source": [
    "# load tcga marker list and metadata\n",
    "#tcga_marker_list = []\n",
    "#with open(\"./kmeans_mask_k2_probes.txt\", \"r\") as fin:\n",
    "#    for line in fin.readlines():\n",
    "#        tcga_marker_list.append(line.strip())\n",
    "\n",
    "df_feat_kmean_mask_k2 = pd.read_csv(open(\"./kmeans_mask_k2_probes.txt\", \"r\"))\n",
    "print(df_feat_kmean_mask_k2.head())\n",
    "tcga_marker_list = list(df_feat_kmean_mask_k2.iloc[:,0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n_probes: 86\n",
      "    chrm     start       end  cluster                     name\n",
      "5  chr17  48593145  48593993        0  chr17:48593145-48593993\n",
      "6  chr17  48593145  48593993        0  chr17:48593145-48593993\n",
      "\n",
      "\n",
      "    chrm     start       end  cluster                    name\n",
      "76  chr2  31233763  31234264        1  chr2:31233763-31234264\n",
      "79  chr2  31233763  31234264        1  chr2:31233763-31234264\n",
      "\n",
      "\n",
      "    chrm      start        end  cluster                      name\n",
      "70  chr6  132400991  132401785        1  chr6:132400991-132401785\n",
      "71  chr6  132400991  132401785        1  chr6:132400991-132401785\n",
      "\n",
      "\n",
      "regions: 85\n",
      "regions minus dups: 82\n",
      "n missing: 1\n"
     ]
    }
   ],
   "source": [
    "# get regions containing tcga probes of interest\n",
    "print(f\"n_probes: {len(tcga_marker_list)}\")\n",
    "regions_of_interest = []\n",
    "n_missing = 0\n",
    "for prb, cluster in zip(tcga_marker_list, df_feat_kmean_mask_k2[\"cluster\"]):\n",
    "    pos = df_probe_metadata.loc[prb,\"Start\"]\n",
    "    chrm = df_probe_metadata.loc[prb,\"Chromosome\"]\n",
    "    found = False\n",
    "    # if the probe is in a region in the 434 set, use that region\n",
    "    for reg434 in modeling_regions[chrm]:\n",
    "        if reg434[0] < pos-1 < reg434[1]:\n",
    "            regions_of_interest.append([chrm, reg434[0], reg434[1], cluster])\n",
    "            found = True\n",
    "            break\n",
    "    # else look for probe in 16K regions:\n",
    "    if not found:\n",
    "        for reg in v1_16K_regions[chrm]:\n",
    "            # regstart < pos < regend\n",
    "            if reg[0] < pos-1 < reg[1]:\n",
    "                regions_of_interest.append([chrm, reg[0], reg[1], cluster])\n",
    "                found = True\n",
    "                break\n",
    "    # else missing\n",
    "    if not found:\n",
    "        n_missing += 1\n",
    "\n",
    "df_regions_of_interest = pd.DataFrame(regions_of_interest)\n",
    "df_regions_of_interest.columns = [\"chrm\",\"start\",\"end\",\"cluster\"]\n",
    "df_regions_of_interest[\"name\"] = df_regions_of_interest[\"chrm\"] + \":\" + df_regions_of_interest[\"start\"].apply(str) + \"-\" + df_regions_of_interest[\"end\"].apply(str)\n",
    "\n",
    "#check for regions assigned to more than one cluster\n",
    "for i, df in df_regions_of_interest.groupby(\"name\"): \n",
    "    if len(df) > 1: \n",
    "        print(df)\n",
    "        print(\"\\n\")\n",
    "        \n",
    "print(f\"regions: {len(df_regions_of_interest)}\")\n",
    "df_regions_of_interest = df_regions_of_interest.drop_duplicates()\n",
    "print(f\"regions minus dups: {len(df_regions_of_interest)}\")\n",
    "print(f\"n missing: {n_missing}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create bed file for regions of interest\n",
    "def make_bed_and_metadata(df_regions, out_pfx:str):\n",
    "    list_of_region_dicts = []\n",
    "    for i, row in df_regions.iterrows():\n",
    "        chrm, start, stop, cluster = row[0], row[1], row[2], row[3]\n",
    "        #chrm, start, stop, genes = region\n",
    "        list_of_region_dicts.append( {\"chrm\":chrm,\n",
    "                                      \"start\":start,\n",
    "                                      \"stop\":stop,\n",
    "                                      \"cluster\":cluster})\n",
    "                                      #\"genes\":genes})\n",
    "\n",
    "    thresh_dicts = [{f\"thresh{n}\": n} for n in range(3,34)]    \n",
    "    calc_dicts = [ {\"col1\":\"total_frags\"}, {\"col2\":\"methyl_frags\"}]\n",
    "\n",
    "    np.savez_compressed(f'./{out_pfx}_col_metadata.npz', np.array(list_of_region_dicts), np.array(thresh_dicts), np.array(calc_dicts))\n",
    "\n",
    "    # write bed file\n",
    "    #df_regions = pd.DataFrame(list_of_region_dicts)\n",
    "    #print(df_regions.head())\n",
    "    df_regions.to_csv(f\"./{out_pfx}.bed\", sep=\"\\t\", header=None, index=None)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create feature metadata for regions of interest\n",
    "make_bed_and_metadata(df_regions_of_interest, \"tcga_k2_clust_regions_v2\")\n",
    "\n",
    "# gsutil cp tcga_k2_clust_regions.bed gs://reference-datafiles/beds/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        0          1          2\n",
      "0   chr13   22848547   22848667\n",
      "1    chr1  109687545  109688046\n",
      "2   chr13  110475647  110475815\n",
      "3   chr17   43953039   43953159\n",
      "4    chr4   13544188   13544834\n",
      "..    ...        ...        ...\n",
      "76   chr5   14405578   14405698\n",
      "77   chr5   83471345   83472266\n",
      "78  chr22   45009678   45010281\n",
      "79   chr5  123088304  123088424\n",
      "80   chr2  179861032  179861756\n",
      "\n",
      "[78 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# create feature using summarizer\n",
    "'''\n",
    "ftio create feature\n",
    "name: hmfc_tcga_cluster_regions_k2\n",
    "numpy_dtype: int64\n",
    "shape: (78,31,2,)\n",
    "column_metadata: /Users/billiegould/Desktop/my_repo/startup/TCGA_COAD_READ/tcga_k2_clust_regions_col_metadata.npz\n",
    "\n",
    "ftio update feature\n",
    "name: hmfc_tcga_cluster_regions_k2\n",
    "version: 2\n",
    "numpy_dtype: int64\n",
    "shape: (85,31,2,)\n",
    "column_metadata: /Users/billiegould/Desktop/my_repo/startup/TCGA_COAD_READ/tcga_k2_clust_regions_v2_col_metadata.npz\n",
    "\n",
    "source ./create_tcga_hmfc_k2.sh\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# methyl cluster aware model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pineapple'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-189-f6856c1bfecf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m# https://github.com/freenome/pineapple/blob/master/pineapple/contrib/components/transformers/methyl/poisson_outlier.py\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mpineapple\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontrib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomponents\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdata_factories\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmatrix_factory\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mMatrixFactory\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pineapple'"
     ]
    }
   ],
   "source": [
    "# import the hmfc model\n",
    "# an example of a pineapple experiment running the methyl model:\n",
    "# https://github.com/freenome/pineapple/blob/master/pineapple/experiments/dc1/methyl/methyl_poisson_spec_rrbs_mask.yaml\n",
    "\n",
    "# transformers used to run the hmfc model\n",
    "# https://github.com/freenome/pineapple/blob/master/pineapple/contrib/components/transformers/methyl/poisson_outlier.py\n",
    "\n",
    "# how to use matrixfactory in pineapple\n",
    "# https://github.com/freenome/research/blob/master/docs/data_access_tutorial.ipynb\n",
    "\n",
    "# from here fly to Jupyter Hub !!! . . . . ."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
